### ISSUE TITLE
security hole in eval()

### ISSUE THREAD

**Issue by ec0ddb47-7b0c-4330-995e-4b03f52fda02** [python/cpython#86638]
BPO | [42472](https://bugs.python.org/issue42472)
--- | :---
Nosy | @vstinner, @tiran

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = <Date 2020-11-26.15:01:19.429>
created_at = <Date 2020-11-26.11:31:04.380>
labels = ['type-security', '3.7', 'invalid']
title = 'security hole in eval()'
updated_at = <Date 2020-11-26.15:01:19.428>
user = 'https://bugs.python.org/cryptophoto'
```

bugs.python.org fields:
```python
activity = <Date 2020-11-26.15:01:19.428>
actor = 'vstinner'
assignee = 'none'
closed = True
closed_date = <Date 2020-11-26.15:01:19.429>
closer = 'vstinner'
components = []
creation = <Date 2020-11-26.11:31:04.380>
creator = 'cryptophoto'
dependencies = []
files = []
hgrepos = []
issue_num = 42472
keywords = []
message_count = 6.0
messages = ['381892', '381894', '381895', '381898', '381900', '381903']
nosy_count = 3.0
nosy_names = ['vstinner', 'christian.heimes', 'cryptophoto']
pr_nums = []
priority = 'normal'
resolution = 'not a bug'
stage = 'resolved'
status = 'closed'
superseder = None
type = 'security'
url = 'https://bugs.python.org/issue42472'
versions = ['Python 3.7']
```

</p></details>

**Comment by divad1196** [comment]
Hi, 

I do not agree with the arguments provided by the author of the issue.
But I wonder if a feature could be added to have a "restricted execution mode" in python directly.
Is it just something considered "too hard", or "not worth" ?

For comparison:
* Ruby as [different safe levels](https://ruby-doc.com/docs/ProgrammingRuby/html/taint.html)
   Nothing similar found
* Some libraries/programs have tried to address this issue, for example [Odoo](https://github.com/odoo/odoo/blob/16.0/odoo/tools/safe_eval.py)
* V8 javascript interpreter does not expose external features.
* Lua is famous as embedded trusted language and can be tweaked according to our needs.
* Compiled languages could use seccomp-bpf to limit their capabilities (at least on linux, and I think something similar is available on Windows)
   There is a [port for python](https://github.com/seveas/python-prctl) but not sure what it is worth.

I think the 2 last points may be the most important ones:
* if could remove most, if not all capabilities, of the program after the startup and maybe some initial imports ? Using Process class with reduce capabilities would be a nice way of protecting the app, etc... 
* if Cpython could be decoupled in "layers", the first one not having access to any IO nor os module, this would be great but certainly this would need a big refactor.


To sum it up: I think there are/may be many ways of addressing this sandboxing issue directly in python with different level of complexity. Any insight about it would be great.
Thank you in advance for your responses.
